### PR TITLE
Move no_grad() to WorkerTemplate._update_all_metrics()

### PR DESCRIPTION
**Pull request by amjltc295**
## Why do we need this PR?
* If no_grad() is not added in the _update() function of a metric, there would be GPU memory leaking issue. 

## How is it implemented?
* As no_grad() is required for all metrics, it is now moved to WorkerTemplate._update_all_metrics().

## Other changes
* Add comments for BaseMetric

### CODE REVIEW

File: src/model/metric.py [hunk 1]
```diff
@@ -21,14 +21,24 @@ def __init__(self, output_key, target_key, nickname):
...
     def update(self, data, output):
+        """ Update metric values in each batch.
+
+        This function would be called inside torch.no_grad() in WorkerTemplate._update_all_metrics()
+        """
         pass
 
     @abstractmethod
     def finalize(self):
+        """ Calculate the final metric values given the variables updated in each batch. """
```
Comment (Nash2325138) [inline src/model/metric.py]: Thanks for the documentation :D
